contracts/src/lib/LibGateway.sol

```diff
@@ -416,6 +416,100 @@ library LibGateway {
+    ///      message execution. In bottom-up, the execution of cross net messages happens in 
+    ///      the parent, which most likely, we have no access to its vm execution 
+    /// @param arrivingFrom - the immediate subnet from which this message is arriving
+    /// @param crossMsg - the cross message to be executed
+    function applyMsgWithRet(SubnetID memory arrivingFrom, IpcEnvelope memory crossMsg) internal returns(bool hasReturn, ApplyMsgRet memory applyRet) {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+
+        if (crossMsg.to.subnetId.isEmpty()) {
+            applyRet = ApplyMsgRet({
+                crossMsg: crossMsg,
+                success: false,
+                ret: abi.encodeWithSelector(InvalidCrossMsgDstSubnet.selector)
+            });
+            return (true, applyRet);
+        }
+
+        // The first thing we do is to find out the directionality of this message and act accordingly,
+        // increasing the applied nonces conveniently.
+        // slither-disable-next-line uninitialized-local
+        SupplySource memory supplySource;
+        IPCMsgType applyType = crossMsg.applyType(s.networkName);
+        if (applyType == IPCMsgType.BottomUp) {
+            // Load the subnet this message is coming from. Ensure that it exists and that the nonce expectation is met.
+            (bool registered, Subnet storage subnet) = LibGateway.getSubnet(arrivingFrom);
+            if (!registered) {
+                // this means the subnet that sent the bottom up message is not registered,
+                // we cannot send the receipt back as top down because the subnet is not registered
+                // we ignore this message for as it's not valid, and it may be someone trying to forge it.
+                return (false, applyRet);
+            }
+            if (subnet.appliedBottomUpNonce != crossMsg.nonce) {
+                return (true, ApplyMsgRet({
+                    crossMsg: crossMsg,
+                    success: false,
+                    ret: abi.encodeWithSelector(InvalidCrossMsgNonce.selector)
+                }));
+            }
+            subnet.appliedBottomUpNonce += 1;
+
+            // The value carried in bottom-up messages needs to be treated according to the supply source
+            // configuration of the subnet.
+            supplySource = SubnetActorGetterFacet(subnet.id.getActor()).supplySource();
+        } else if (applyType == IPCMsgType.TopDown) {
+            // Note: there is no need to load the subnet, as a top-down application means that _we_ are the subnet.
+            if (s.appliedTopDownNonce != crossMsg.nonce) {
+                return (true, ApplyMsgRet({
+                    crossMsg: crossMsg,
+                    success: false,
+                    ret: abi.encodeWithSelector(InvalidCrossMsgNonce.selector)
+                }));
+            }
+            s.appliedTopDownNonce += 1;
```

Comment:
It's still not clear to me what these "vm" errors that we're supposed to handle by doing things in the host should be, and how they should work for example with this nonce. 

For example I see that we adjust the nonce, then execute, returning whether the Solidity sub-transaction was successful or not. This is good, the next message can be handled and it should match the next nonce, all is well. However, what happens if we encounter this "vm" error? I assume it means the nonce was _not_ executed in those cases, so even if the host sends the next message, it will fail the nonce check, no?